scipy/linalg/_decomp_lu.py

```diff
@@ -174,53 +182,166 @@ def lu(a, permute_l=False, overwrite_a=False, check_finite=True):
         Whether to check that the input matrix contains only finite numbers.
         Disabling may give a performance gain, but may result in problems
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
+    matrix_p : bool, optional
+        If ``False`` the permutation information is returned as row indices.
+        The default is ``True`` for backwards-compatibility reasons.
 
     Returns
     -------
-    **(If permute_l == False)**
+    **(If `permute_l` is ``False``)**
 
-    p : (M, M) ndarray
-        Permutation matrix
-    l : (M, K) ndarray
-        Lower triangular or trapezoidal matrix with unit diagonal.
-        K = min(M, N)
-    u : (K, N) ndarray
-        Upper triangular or trapezoidal matrix
+    p : (..., M, M) ndarray
+        Permutation arrays or vectors depending on `matrix_p`
```

Comment:
Just  to bikeshed a bit more:
1. Is it OK to call this `matrix_p` when now this can be a "stack of matrices" vs "stack of 1d arrays"? Probably yes, but we were already discussing the shade of the shed. But for a `(q, M, M)` shaped input and `matrix_p=False` we get shape `(q, M)` for `p`, right? Which would arguably "a matrix" as much as the `(M, M)` shaped original behaviour. Again, it's clear to me, and probably to most people, just want to make sure this is the least bad choice for the name.
2. Is it OK to keep the `(..., M, M)` shape in the docs here? Is there a way to spell this out correctly (or close enough) without building a 2x2 block of "if `thing` is `bool instane`" matrix?

Reply:
Yes, the ndarray case is a generalization of 2D case hence storyline follows the 2D case and where available the sizes are replaced with it. I am practically following what NumPy does,  say with, https://numpy.org/doc/stable/reference/generated/numpy.linalg.qr.html . So the idea is that the returning thing is a lot of P matrices or a lot of P vectors in the ndarray case. 

But admittedly, story relies a lot on the legacy stuff. And to be honest, I don't know what the actual story line should be after looking at this line for a very long time.

Reply:
Just to be clear, my second point is that you have `(..., M, M)` whereas the linked NumPy `qr` has
```
‘raw’ : returns h, tau with dimensions (…, N, M), (…, K,)
```
which is exactly the kind of thing I'd expect for the new feature :)

Reply:
Oops, no I think I misread the above line. Let me reread...

Reply:
OK, [`numpy.linalg.svd`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html)
```
    full_matrices : bool, optional
        If True (default), `u` and `vh` have the shapes ``(..., M, M)`` and
        ``(..., N, N)``, respectively.  Otherwise, the shapes are
        ``(..., M, K)`` and ``(..., K, N)``, respectively, where
        ``K = min(M, N)``.
```
So it spells out the two shapes with the flag, and then with the return value
```
    U : { (..., M, M), (..., M, K) } array
        Unitary array(s). The first ``a.ndim - 2`` dimensions have the same
        size as those of the input `a`. The size of the last two dimensions
        depends on the value of `full_matrices`. Only returned when
        `compute_uv` is True.
```
I think _this_ is indeed similar.